src/templater.rs

```diff
@@ -495,18 +496,30 @@ impl Template<()> for CommitOrChangeId<'_> {
     }
 }
 
-fn highlight_shortest_prefix(id: &CommitOrChangeId, total_len: usize) -> String {
-    let prefix_len = id.repo.shortest_unique_id_prefix_len(id.as_bytes());
-    let mut hex = id.hex();
-    if prefix_len < total_len - 2 {
-        format!(
-            "{}[{}]",
-            &hex[0..prefix_len],
-            &hex[prefix_len..total_len - 2]
-        )
+/// This function supports short `total_len` by ensuring that the entire
+/// unique prefix is always printed
+fn extract_entire_prefix_and_trimmed_tail(
+    s: String,
+    prefix_len: usize,
+    total_len: usize,
+) -> (String, String) {
+    let mut prefix = s;
+    // Prefix is not limited by `total_len`
+    let mut rest = prefix.split_off(min(prefix_len, prefix.len()));
+    rest.truncate(total_len - min(total_len, prefix_len));
+    (prefix, rest)
+}
```

Comment:
I tried following yuja's example of reducing allocations. I haven't actually checked if I removed any but, in any case, it sounds like I overdid it.

I'll rewrite it.

Reply:
I do want to extract this function so that I can reuse it later. I think I'll need to add some lifetimes so that the function takes in a slice and returns two slices.

Reply:
fwiw, I don't think prefix_len should exceed the total_len. For `commit_id.short().whatever_style()`, I would expect the id would be clamped by the `short()` constraint.

Reply:
I want to be able to set `total_length` to, say, 6 for the `jj` repo to save space. At the same time, I want to trust that if a rare id *needs* to be 7 characters, `jj` will show me all 7.

Reply:
I think this should work:
```
fn extract_entire_prefix_and_trimmed_tail(
    hex: &str,
    prefix_len: usize,
    total_len: usize,
) -> (&str, &str) {
  let prefix_len = min(prefix_len, hex.len());
  (&hex[0..prefix_len], &hex[prefix_len..total_len - 2])
}
```

Reply:
@yuja, I think you're saying you want the behavior of hg's `short(node)`, while @ilyagr wants the behavior of hg's `shortest(node, 6)`. Maybe we should rename this method to `shortest()` and add a `short()` with the behavior you want?

Reply:
> Maybe we should rename this method to shortest() and add a short() with the behavior you want?

That sounds better. Actually I'm fine with either behavior, but I really don't like the inconsistency that `.short()` clamps at 12 chars, but the others don't.

Reply:
I changed the function to a minor variation of Martin's suggestion above. It was a surprise to me that Rust panics when the end of the slice is before the beginning, but is fine with endpoints being larger than the length.

Let me know if we don't need the tests I added to convince myself.

The main problem with making `short` aware of the prefix length is that it's expensive to compute prefix lengths. I'll work on the renames you suggested later.

Reply:
Since `FormattedCommitOrChangeId` holds `repo`, the shortest length could be calculated when `.shortest()` method is called, not when the template is rendered.

Perhaps, the struct can keep track of `trim_display_to: usize, shortest_len: Option<u32>, bracket: bool`. And,

* `short()`: set `trim_display_to = min(12, trim_display_to)`
* `shortest(min_len = 12)`: set `shortest_len = ...`, `trim_display_to = min(max(shortest_len, min_len), trim_display_to)`
* `bracket()`: set `bracket = true`, subract `trim_display_to`?

On rendering, prefix/rest can be labeled if `shortest_len` had been calculated.

Perhaps, in future, `bracket()` will be replaced with a template like `apply(commit_id.shortest(10), prefix "[" rest "]")`.

Reply:
> It was a surprise to me that Rust panics when the end of the slice is before the beginning, but is fine with endpoints being larger than the length.

It is not supposed to be fine:

```rust
fn main() {
    let _ = &b"foobar"[6..12];
}
```

```txt
thread 'main' panicked at 'range end index 12 out of range for slice of length 6', t.rs:2:14
```

Reply:
> Since `FormattedCommitOrChangeId` holds `repo`, the shortest length could be calculated when `.shortest()` method is called, not when the template is rendered.

And maybe it's even better to add a separate `ShortestCommitOrChangeId` struct?

* `CommitOrChangeId::short()` -> String
* `CommitOrChangeId::shortest(min_len)` -> `ShortestCommitOrChangeId { prefix_len, total_len }` or `{ prefix_len, truncated_hex }`
* `ShortestCommitOrChangeId::brackets()` -> String

It should avoid inconsistent evaluation of `trim_display_to` value.

Reply:
@samueltardieu : That's right. What I learned on top of that is that `b"0123456789"[10..10]` is A-OK.

@yuja: I think all I'll do in this PR is rename appropriate things `{short=>shortest}`. We can leave the rest for later. At least, I don't feel like tackling it today.

I added a few first thoughts about your suggestions below. I don't feel like I thought everything through completely yet. For example, I haven't decided what's important to get right and what is just bike-shedding when the possible decisions are about equal.

> And maybe it's even better to add a separate ShortestCommitOrChangeId struct?

I was vaguely thinking of separating the highlighting and the trimming entirely, to something like `change_id.styled_prefix().shortest(12)`. Or, perhaps `styled_prefix` would be the default, and then you'd have to do `change_id.plain().short(12)` to disable prefixes.

For trimming, a string-valued `change_id.plain()` makes sense to me.I haven't formed an opinion on whether a whole separate struct is worth it.

> Perhaps, the struct can keep track of `trim_display_to: usize, shortest_len: Option<usize>, bracket: bool`

This sounds pretty good if it's really important to hard trim the commit id to some length. As I was saying before, I couldn't think of a good reason to print an id if its unique prefix isn't included. I don't feel strongly about this, though.

One possible change I would make to this is to somehow make it clearer what determines whether the performance hit for calculating unique prefixes happens or not. I'm guessing, in your exact suggestion, `shortest_len` being `None` or not determines that. It would need some documenting, I wonder if it could be made more self-evident.

> Perhaps, in future, bracket() will be replaced with a template like apply(commit_id.shortest(10), prefix "[" rest "]").

That sounds great. I'm not sure if it's worth implementing just for the brackets, but I think it will work nicely with similar operations on objects other than `commit_id`.

I wouldn't be opposed even to `commit_id.shortest(10).prefix "[" commit_id.shortest(10).rest "]"`, though that does raise the question of introducing variables to the templating language.

Reply:
I did the renaming, so I'm marking this resolved. The rest of discussion should of course continue.

Reply:
> I was vaguely thinking of separating the highlighting and the trimming entirely, to something like change_id.styled_prefix().shortest(12). Or, perhaps styled_prefix would be the default, and then you'd have to do change_id.plain().short(12) to disable prefixes.

IMO, labels are just hints, and the styled shortest string are equivalent to the unstyled one. That's why I suggested `change_id.shortest(min_len)`.

> I haven't formed an opinion on whether a whole separate struct is worth it.

It will eliminate the whole problems of method combinations. If `commit_id.short().shortest_prefix_and_brackets()` weren't allowed, we wouldn't have to think about that.